### PR TITLE
DAO support for Coverage Pool underwriters

### PR DESCRIPTION
**Pull request by pdyraga**
~~Depends on #202~~
~~Marking this PR as a draft until ^ is merged.~~

In #199 we added support for DAO `Checkpoints` to `UnderwriterToken` contract (COV token). This is not enough to enable DAO support in coverage pools - we know one's voting power proportional to COV supply but T DAO needs to know one's voting power proportional to T supply.

To make it happen, `CoveragePool.getPastVotes(address account, uint256 blockNumber)` implemented in this PR computes underwriter's voting power as underwriter's coverage pool share multiplied by the amount of T coverage pool has. This approach takes into account liquidations and rewards emitted to the coverage pool.

To participate in DAO, the underwriter needs to delegate their COV voting power to someone (or self-delegate), just like in the case of T liquid holders.

### Example 1

- Underwriter 1 deposited 400 T into the coverage pool.
- Underwriter 2 deposited 200 T into the coverage pool.
- Coverage Pool has 600 T.

Underwriter 1 voting power is 400, Underwriter 2 voting power is 200.

### Example 2

- Underwriter 1 deposited 400 T into the coverage pool.
- Underwriter 2 deposited 200 T into the coverage pool.
- Coverage pool has 600 T.
- Underwriter 1 withdrew 100 T from the Coverage Pool.

Underwriter 1 voting power is 300, Underwriter 2 voting power is 200.

### Example 3

- Underwriter 1 deposited 400 T into the coverage pool.
- Underwriter 2 deposited 200 T into the coverage pool.
- 90 T was allocated to the coverage pool as rewards.
- Coverage Pool has 690 T.

Underwriter 1 voting power is 460, Underwriter 2 voting power is 230.

### Example 4

- Underwriter 1 deposited 400 T into the coverage pool.
- Underwriter 2 deposited 200 T into the coverage pool.
- 60 T was liquidated from the coverage pool by the risk manager.
- Coverage pool has 540 T.

Underwriter 1 voting power is 360, Underwriter 2 voting power is 180.

### CODE REVIEW

**Review by cygnusv — APPROVED** [review]
LGTM! One thing that I noticed is that tests can be improved a bit wrt to different delegation possibilities of the underwriter token.
Also note that the Checkpoints library was changed in the Threshold repo, although it's a minimal change on contracts dependencies. I mention this in case you need to bump some of your dependencies here.

**Comment by dimpar** [comment]
No more comments from my side. Looks good.

File: test/CoveragePool.test.js [hunk 1]
```diff
@@ -431,4 +447,319 @@ describe("CoveragePool", () => {
...
+        ).to.equal(0)
+      })
+    })
+
+    context("when single underwriter deposited into the pool", () => {
+      const depositedAmount = to1e18(401)
+
+      beforeEach(async () => {
+        // Mint to some unrelated account
+        await collateralToken.mint(thirdParty.address, to1e18(100000))
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply test/CoveragePool.test.js:527]: We do it to ensure the total supply of T that exists in circulation, outside of the coverage pool deposits does not affect the voting power of underwriters.

Clarified in https://github.com/keep-network/coverage-pools/pull/206/commits/a0901c8954ce700167acd43fc3577bacc12b9221.

File: contracts/CoveragePool.sol [hunk 2]
```diff
@@ -252,6 +254,39 @@ contract CoveragePool is Ownable {
...
+    function getPastVotes(address account, uint256 blockNumber)
+        external
+        view
+        returns (uint96)
+    {
+        uint96 underwriterVotes = underwriterToken.getPastVotes(
+            account,
+            blockNumber
+        );
+        uint96 underwriterSupply = underwriterToken.getPastTotalSupply(
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/CoveragePool.sol]: Good point, done in https://github.com/keep-network/coverage-pools/pull/206/commits/03cacef02b34af8744282a9ca2a5ea328437690d.

File: contracts/CoveragePool.sol [hunk 3]
```diff
@@ -252,6 +254,39 @@ contract CoveragePool is Ownable {
             );
     }
 
+    /// @notice Determine the prior number of DAO votes for the given coverage
+    ///         pool underwriter.
+    /// @param account The underwriter address to check
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/CoveragePool.sol:259]: Good idea, added in https://github.com/keep-network/coverage-pools/pull/206/commits/a0901c8954ce700167acd43fc3577bacc12b9221. I have also fixed the non-delegated case in https://github.com/keep-network/coverage-pools/pull/206/commits/8b464f4b27eba35b5b7c9d1f83cc9adee5939fa2.

**Review by dimpar — APPROVED** [review]

**Comment by pdyraga** [comment]
@cygnusv Please take a look when you'll have a chance. If you give 👍🏻 after the latest changes, we should be ready to go.

File: test/CoveragePool.test.js [hunk 4]
```diff
@@ -477,20 +477,19 @@ describe("CoveragePool", () => {
...
       })
 
       it("should return zero", async () => {
         expect(
           await collateralToken.getPastVotes(
-            underwriter1.address,
+            thirdParty.address,
             lastFinalizedBlock
           )
         ).to.equal(0)
```
Comment (cygnusv) [inline test/CoveragePool.test.js:495]: @pdyraga What's the purpose of this test? `underwriter1` (now `thirdParty`) was minting 400 tokens, approving the asset pool and depositing all of them there; after this, we check that they have 0 votes, but this is going to happen always, regardless of them having delegated voting or not, because they have 0 tokens in the end anyway.

Comment (pdyraga) [reply test/CoveragePool.test.js:495]: >  but this is going to happen always, regardless of them having delegated voting or not, because they have 0 tokens in the end anyway.

I am not following this part. This is how I understand this test:

Deployer minted 400 tokens to `thirdParty`. `thirdParty` deposited 400 tokens to the coverage pool and received COV tokens. We check that the `thirdParty` has 0 voting power because they have not delegated this voting power to anyone - including self - even though they have 400 COV.

In other words, this test ensures that depositing to the coverage pool is not enough to receive voting power. One must delegate voting power on the COV they have.

Comment (pdyraga) [reply test/CoveragePool.test.js:495]: tl;dr; @cygnusv was right 🙈 https://github.com/keep-network/coverage-pools/pull/206/commits/31d3af2039adc8a556e2ed125b28af4a7a16a371

**Comment by pdyraga** [comment]
@cygnusv @dimpar This is ready for another chance.

**Review by cygnusv — APPROVED** [review]

**Review by dimpar — APPROVED** [review]
